### Google PlayStore Data Cleaning

In [1]:
import numpy as np
import pandas as pd

#### Load Dataset

In [2]:
raw_data = pd.read_csv('googleplaystore.csv')

In [6]:
raw_data.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

  Reviews Size  Installs  Type Price Content Rating  \
0     159  19M   10,000+  Free     0       Everyone   
1     967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [9]:
# Make the copy of the original data
df = raw_data.copy()

In [8]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

  Reviews Size  Installs  Type Price Content Rating  \
0     159  19M   10,000+  Free     0       Everyone   
1     967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [11]:

df['Rating'].unique()

array([ 4.1,  3.9,  4.7,  4.5,  4.3,  4.4,  3.8,  4.2,  4.6,  3.2,  4. ,
        nan,  4.8,  4.9,  3.6,  3.7,  3.3,  3.4,  3.5,  3.1,  5. ,  2.6,
        3. ,  1.9,  2.5,  2.8,  2.7,  1. ,  2.9,  2.3,  2.2,  1.7,  2. ,
        1.8,  2.4,  1.6,  2.1,  1.4,  1.5,  1.2, 19. ])

In [14]:
# Fill the nan of Rating column with the median of that column
df['Rating'] = df['Rating'].fillna(df['Rating'].median())

In [17]:
df["Rating"].isnull().sum()

0

#### Observation:
There is no missing value in the Rating column

In [21]:
df.shape

(10841, 13)

In [19]:
# Let's see the another column
df['Reviews'].str.isnumeric().sum()

10840

#### Observation:
Since there are 10841 rows but there is only one row with column non-numeric type data 

In [20]:
df[~df['Reviews'].str.isnumeric()]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [23]:
# Drop the column with type non-numeric in Reviews 
df = df.drop(df.index[10472])

In [24]:
df.shape

(10840, 13)

In [29]:
df[~df['Reviews'].str.isnumeric()].sum()

App               0.0
Category          0.0
Rating            0.0
Reviews           0.0
Size              0.0
Installs          0.0
Type              0.0
Price             0.0
Content Rating    0.0
Genres            0.0
Last Updated      0.0
Current Ver       0.0
Android Ver       0.0
dtype: float64

#### Convert the object type of column "Reviews" to int type

In [34]:
df['Reviews'] = df['Reviews'].astype(int)

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          10840 non-null  float64
 3   Reviews         10840 non-null  int32  
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), int32(1), object(11)
memory usage: 1.1+ MB


In [36]:
df['Size'].unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

In [43]:
df['Size'] = df['Size'].str.replace('M','000')
df['Size'] = df['Size'].str.replace('K', '')
df['Size'] = df['Size'].replace('Varies with device', np.nan)

In [59]:
#
df['Size'].mean()

TypeError: Could not convert 19000140008.7000250002.80005.60001900029000330003.100028000120002000021000370002.70005.500017000390003100014000120004.20007.0000230006.0000250006.10004.60004.20009.20005.200011000110004.20009.200024000110009.400015000100001.20001200024000260008.00007.90002500056000570003500033000330005.600054000201k3.60005.7000170008.60002.4000270002.70002.50007.0000350001600016000170003.40008.90003.90002.9000380003200037000150005.400018000380001.10007.9000350001700019000140002.20004.500017000140009.8000210005200014000250009.000012000350006.7000300005.70002.9000170002.90002.60004.20007.1000570003.700022000240007.4000210003.40002.90003.10006.40003.20008.20009.90002.9000230004.60003.10004.90009.50004.20005.40002.80005.00005.9000130007.10006.700017000190006.7000210002.70003700015000230001900023000730004.90006.80002.90003.50004.0000210002.30007.2000100006.10002.1000420007.30003000029000350009.1000250003.9000180001200021000550003.200023k16000140003700011000250007.30006.5000250003.10001.50007.50008.60001.20003900014000190006.80003900014000200002600051000410002000012000480001000022000110008.50008.60002800028000370009.00004600014000260002300014000200002600051000410002000012000480001000022000110008.50008.6000280003900014000190006.8000390003500029000410004.90002700032000120001500011000280002.20003.40008.30004.30004.700010000150007.1000210006.10001500011000290003.30002200040000100009.10004.7000380006.7000370007.8000190005.700035000170008.8000160001100011000150006.60006.60005.1000180004.000037000220003.3000320003700017000350004000061000660001100079k250001400024000170008.30008.20008.40004.0000320006.10002.80003.3000400002.20004.30004.70002.30002.3000110002.700014000130003.7000100001300010000130008.80005.50002000029000110001700017000118k160005.1000440007.300030000695k1.6000190001100023000260006.20002000018k1.200017000150005600012000110002900040000530003.100024000230005.000044000270006.10002100011000210002400031000270006.2000120008.00005.90007.90005.00001.4000130004000013000190005.000019000270003.0000130007.2000250005.70005.50006.50005.80003.8000410007.90002.80002000015000280009.600013000150004000056000120009.40002900011000190002300027000190004500021000400002500015000240001100038000310006.1000270003200013000630004400016000280005.9000130009.0000200007.900028000490002700018000140002700041000380005.5000270007.2000210008.800038000110006.500026000190006.1000770009.50001400038000160003.40004.70005.00004.90004.9000160003.90004.40004.80004.5000700004.90003.70004.9000160004.60003.600021000130008.70009.20006.9000390008.00003.90001400011000130009.30005.000010.00005.00005.00003.6000190005.00009.8000140005.00004.90005.5000110003.80005.00008.20002100012000100003.30002500015000100004.90004.40003.50007.80009.000015000190005600038000160005.0000270004.70003.40004.900025000290008.100014000270003.9000130005.00005.5000160004.8000140004.70003.70004.90003.50004.5000360004.5000190001400077000210007.90009.30001800018000210003.300024000390003.20005.100011000270003700026000110004100049000210008.10005100014000180003.000019000220006.90007.40008400025000180002.50003.900021000270002100018000180001000097000170002.00001.90001.80001.800017000180005.30005.40005.40005.4000150005.3000480005.400047000556k290002.30004.40006.6000526k12000290002100018000180001700010000760001700018000210006.9000140007600011000190006.50007.00003.3000210002.6000210005.20001800016000150001.2000180001.80007.6000590006.900014000410001900076000210002100021000210005.40007.3000590004100021000760007.600013000630004400024000240004.5000110004.60003.30006.5000120005.60009.700015000520004.50004900018000130004.0000160001700024000120007800025000570009.100039000160008.5000120007200012000160009.600022000130001200023000200002500017000190002100019000190004.4000190001200011000120002500021000130001700020000200001900017000210001500019000190007.20003200044000190001100025000160002000021000300005700019000170001500019000190001900044000190003500044000200001900025000430003.60007.7000440001200053000770009.5000360006.30005.9000120002.8000260008.70001600053000140008.900011000334k24000130002.80009.70001.40003.5000610006.00002.30009.000011000270003.6000290004.400021000130002700011000240006.10006.300023000340003.80004200019000700003200093000400002400020000150002800010000140004200065000370003900047000790001800010000032000460009.80008.20001800039000440005.0000210002200023000270005.0000320007.40002400022000100001700038000460003.80009.30008.4000280009.7000110007.400019000160001400022000250003.60004.5000230004.200058000220001.40009.10002200021000110003800012000450005000024000470002300046000450005000024000330001200053000140006800037000790003200046000400003200024000420001200013000130004.90008.90001900014000640008.400011000660004100015000350003600017000250003500010000120001100017000430004100043000160003900013000300007.10002300017000180001700015000220009.00007.20002.3000270008.200076000220001900028000350002200025000300008.50001700034000340001700016000150002.2000110006.900013000110002.9000250007.0000270007.5000150009.40006.40005.50003.900055000280008.80004.00006.9000110005.5000150004.30007.100058000180001.50006.500011000670005700059000480005700013000310001000057000210002500093000600002700054000590007800055000670003500043000390004.2000440004500028000290001200015000310002000039000940003500041000490002200023000280007.20006.50002400020000150005500038000320004000059000230006.10005800055000570009300060000190003.800018000570005.50005.7000510009.40003.30004.0000190003.80004.3000180005.70003.60005.50005700051000580009.40003.30004.00009.900043000370004000073000280002300010000150008.40003100041000200003900035000340007.5000120007.70002100018000400008.70008.6000180007.50009.10001300015000190005.1000260008.3000270007.9000270005.1000150003.1000160009.2000100002.6000130001.90007.0000150004.20007.8000770005.70005.90007.60005.80005.80005.30005.500025000350007.9000190007.5000120003400030000160001600015000210001.80004.3000100003.7000232k99000624k6.90007.00007.00006.4000160005.9000950008.5k200002.5000500001000018000120002.50002.10002200041k22000292k5.50002.50008.100028000240003.10003.400011k13000190009.5000210009.40002.4000360005500032000340005.1000100000800003.3000280005.000011000550007.10003.20004.4000160006.3000450002300037000130008.1000130001400030000360003.00001.4000390001.80001.7000190001500010.00002.7000160002100020000140007600018000120006.20006.80002400019000130001500013000440009.700024000130004.50004.50002900012000120004.6000100006.800033000120004.6000100006.8000330001300015000230007.700015000250003.0000220007.5000240004400014000670007600074000230004600024000970006200024000330007.80004600069000750006700050000980004.9000370001800052000780003.90005900097000110003800085000780006300024000690007500070000630009900049000760006700024000520009800074000690008200096000990004600099000870003.900067000530007.80009700059000460007500050000620002300072000700001800035000780002500074000970001700053000110006000057000140006700033000760007400040000590005200010000062000850008700033000320004000029000400002400038000520005400071000730005700038000620009500016000390007800036000310009800024000850003300010000036000860008000087000790003000087000910006300095000480009300053000820009900049000250009100052000560005600085000810008200098000770009900096000960008200070000630003500048000840001500069000460001500033000480006700074000570005200025000360003100091000920007000093000520009100037000640003500097000930002800083000990005500082000140008100074000670005200076000670005000046000970009800070000690003.90007.800018000240006200011000870009900082000940009600075000140005900074000290001700097000780006300025000290007200053000760003300060000620007800075000700009900099000270007000088000630008200048000310002300059000390005400048000300006600032000500005300061000490003800020000660005700095000320003300066000750002900074000690007.80003300052000500004600012000410006300053000630009900020000110007500013000460007800049000140001000009700015000330007000010000770004.9000250009600023000510009600070000380002400020000670001900051000220004600021000390002300026000390002400015000200002400044000380005200014000490008.90005600099000100006.90002000010000190003300085000480009.600015000500005100024000130001600063000150002600081000790009400046000430004500091000910005300028000260005600080000600006700078000270006300026000460008300060000440008200067000990002400078000510009400016000930007700032000580004800069000730009900016000540005600068000250007.0000540005100048000440008.700091000970002600016000530009500083000780005600021000260006000022000950004800058000910002200036000370007000034000490009.8000370005500084000160009200057000240006.9000440001100016000170001800016000190001.600047000160001400075000250005.7000170003.5000160004200047000670002600072000370001.2000170001900023000250008300094000970003700092000480001200015000830009.50001900023000250008300094000970003700092000480001200015000830009.500067000630004900014000580002200015000500005900034000540005300037000250002600061000580009100017000830001700064000410005800023000710002300012000704k2.90002500020000210005.40002500042000320001.8000330003.80005.80009.5000190005.000023000862k9.9000240002.4000899k380006.5000140001.200048000378k220004.800025000266k375k620001.20002.8000950002.60004.60005.10001.800053000390005.3000140008.400018000410001000003.4000680002.600011000370009.800023000180007.400011000230003400015000130002400020000140006900038000260002.600021000380001600021000100002200014000360001600061000110008.50003700022000170007.700039000300003.70002800016000200008.900097000120007200021000270005.70006.6000120005.90002000015000260003.300044000320001200036000170005.80001100016000250006.100022000280006.30002500060000200008.0000200006.4000340003.8000420002.4000700005.0000280002100038000740001.800012000130002.90004800019000680006.8000120005.80004100025000420006.5000130001.50003.20007.70001.3000260005.4000110009.500062000180003.10001.8000660005.00002.900022000160001900040000230001500029000160006500029000200003.800019000250005.1000160001.500010.0000220003.60002000020000130002.50003.6000240001000025000290001700053000830004.60004.60002800029000460005.70001.4000120002.20002700018000300003.50004.20006.90006.60007.0000110005.6000290003.3000200007.00003.3000290007.100023000240001400013000340004.300043000690002200020000250002.00004000037000400004.30005.70004.000017000110004.4000150002.3000400008.4000110002.90002.40001500024000290002400022000860002200011000975k980k6.0000200002600013000230002.70004.2000280006.1000150001.2000370005.9000140002.3000320002.70002.2000260004.00001.50003.70002.80003.90003.10002.8000200009.90005.3000100002.6000310002.30006.000017000200006200015000180005.40002.8000150005.30003700023000280003500076000340002.70004.1000850001500089000200009.30007600034000160008.4000200007600068000130005000023000280007.000023000760006800013000560008.40002300022000150003000030000270002000042000180004.200030000330007.90001200015000100009.90009.0000210001300011000210008.1000120007.30001500022000110008.300020000380009.100024000290009800025000180006.100014000210004300015000230005200057000200004000017000130003300018000290003100020000130007.3000120008.800018000140002100043000340001500026000340002.7000520002900020000230004200016000120006.50001500017000240002200015000180008.8000140002.700029000200006.5000150006.200019000270001.10002.7000160008.800012000240005600015000180001400022000280005900037000360009.500082000280009.700022000170007.70001100021000190008.70003000025000170001300023000220009.10008.30001900053000170001000029000460004.50004.9000240002300025000260002.0000190009.6000130004.200016000590009.5000530006.600027000220005000030000170001800010.00009.900023000510003500016000110007400048000130004.200016000590005.60001.50005.70006.10001.90004.0000470002.00009.50002200014000100006.90004400051000480004300021000250003.90004.0000510009.60009.50005000021000310002700053000340001200050000470009.20005100048000240001.60001700031000590004600053000450009.80003400020000190006.0000150006.50006.2000140006.100017000240001300010000100006.900031000280002.600030000130002700029000180003500017000180003.0000320002400010000340006.600027000250006.6000240002500033000880009.3000840004.700025000120009.900061000250005200057000410004.6000100002300027000330004.70002.200095000340009.4000190003.90003400021000190006.60008200019000260005.200016000330006.000025000340008.100035000250001900048000210006.600032000820002300052000140002500058000190002900081000280007.600033000420008.3000140003700014000210002800027000390001700039000510008600026000460002900027000140001000071000150001100062000800005.40009.80007.600026000130006200010.000015000800001400057000510008.300024000290003.1000460005500019000190004000029000430005000013000150004.000014000190002900044000220002800012000460007.6000290004.10008.5000310004.100022000280005.90003.9000150004.30005.3000160007.50003.70005.8000170008.50009.10007.60002.5000150006.100017000580004.6000160001.30002.30001.90002.700016000110005.30004.20009.10004.30001.80008.50004.40003.800011000110007.40005.30005.40003.60004.3000270001.80003.30002.00006.10009.60009.6000100008.200014000250009.9000220008.90002.4000250001.8000170004.10001800014000696k4.70008.10007.90004.6000544k3.5000525k2.1000140008.7000210001200018000170003.80002300014000150009.90001200012000200007.40003.000014000140007.2000220007.200026000380003.90005.40009.80007.60001.10009.10006.8000120003.3000120009.8000920k140006.8000270007.4000290005.90001400014000260001.90003.9000779k5.100010000130004.00005.40003.30004.3000853k210006.40008.60002.10007.10003.30003.50004.10009.70003.30007.10008.40007.00005.5000130005.9000240006.20006.90006.1000240004.100050000110005.10001.30002.30001600015000610001.6000490007.20001.500049000140001.200010.00004.30003.8000320004.50008.40001200025000120007.1000160006.50009.2000720k2.20004.10006000037000713k4.70006.60002.5000190004.200011000120005.70009.00006.90008.40001.90001500011000610004.10003.80003.80002.80008.20003.30006.20008500014000950005.50004200015000980006.40002.80006.50003.60002.40003.40001.50005.00006.6000590006.8000240006.60004.90009.1000110005.8000240004.400016000110004.9000110005300020000380004.900017000370005.200028000710004600016000180001800010000170009.7000210003.200015000440004.800010000540009.10001900012000190007.6000380006.10009.2000110009.200020000220005.30005.60005.40002.900025000230007.20003.3000170006.0000330003.10004.10002700044000130003.00006.100064000470005.400027000500004.000013000440002.5000230002300014000890001600026000110005600025000170009.100011000150009.70005.50006.70002.9000130008.500025000120006.30006.300019000310009.0000250004.60008.0000230009.80008.70001200010000190006.6000140008.60001300013000100007.5000360008.20002200023000140002500027000350008.60004.60002500012000180003.100014000230008.800035000360002500012000230001300013000220004300033000320006.80004.00003.70003.4000480007.700060000600005.60005.2000250003.300029000490003.6000480005.4000180003.2000110008.5000240002600014000780007.00009.20004.90006.500011000960002.90002400055000740002400032000150006.3000970007400014000140005.70001900050000760007000029000160001300073000170008500099000920003700045000570002600047000220005600050000420003400070000170004.100011000110009400024000560006600063000910001700021000810003.000059000830001900023000320003900057000520007.50007.40008.90006.300018000140002000096000350008.40008.600055000290001300016000400006.7000100000940004.4000640004.50006.00007.40003.700025000100002600098000970001.8000110007800050000740008.30007.70002.50005.30004.00004.2000740006.90006.6000160004.70005.40003500026000320002.70003.70007.00006.9000630002.2000300005.80003.10001.50008.6000772k3.30005.00001.10004.40002.10001.600029000390002900014000410009.7000600006200047000890009.900064000480001200061000320002800036000130006900074000180003.60007900066000420009.80002.8000400002000025000440008.8000160008900058000200004.8000318k130004.900012000660005000059000140005.9000410003700022000150006.10003.100058k79000150005500011000130004.90007.60003.0000370009.300023000241k110009.20004.80001600040000130004.900031000140003.00009.90003.70001.700011000630002.10001.4000360003200059000370007.9000250001900039000196k2.30008.1000510003.40008.800028000110006.1000140009400011000140001.6000150002.70005.7000470001.6000857k250001.30001.5000330003.90008700014000100005.9000580001.9000480005700051k4.50008.900026000150001.50002.2000420005.60008.50008.900013000130005.4000270003.0000190008.3000570001.20002.8000190007.20005200065000370002.900011000340006.50003.90003.9000570002.8000230005500075000400005000025000250006.600036000750008.8000430009.900096000953k9.0000230002.600023000350007.2000490004.700043000970003.10006.7000130007.10009.100098000300008.9000130002.4000440007.3000330003.40003.80003.60004000014000340004000011000630001400061000630008.9000100004400010000590007.40002300050000440001500040000620002.10003100019000260003.6000130004.5000350007100015000230003.500060000400007800020000490004000048000870002000050000865k2.9000670002.4000610006.80002.60001300029000840005.40002.80002.90006.900046000270003700082000440007.300037000530004400024000340009.100038000150002600063000251k4.20007.3000490009.70001.40004.90006.20006.00001.200016000590007.7000300006.4000640004600076000540008.400022000230008.20003.8000540009.10004000028000110003400058000210004500022000280001700031000590009100020000690007100013000330003.30006.4000200007.6000620003200049000930k340004900024000720008.7000400006.8000470003.60002.9000260001.300045000540k4.6000313k8.6000746k10000380005.10006.1000280005.2000220003400051000390002.100012000110007.500024000190001.90005.20002.30004.50001.70001.80006.30005.50006.2000280004.30005.70004.90003.60002.0000160007.40004.30007.3000600003.700024000100007.5000130003.60002.0000140002.30001600020000210002.1000260001400018000600007.10006000040000550002000019000170003.60004.50003.60003300043000210002.3000140001900095000290002.7000160003.90003.10001200039000260001900067000120002.80006.20002.90005.9000600004.900062000960009.20002.80003.90002.2000203k370003.70007.70002.900029000340003.60005.7000640003.400058k6.7000230008.9000570001200013000930008300041000280002.70004.40003.4000170002.20007.10002000026k3100045000330001400037000240006.50004000093000314k3.7000570002.700068000250002.4000960002.200033000190002100026000290003.700025000280002.1000150007.9000460008.900013000200002.30004.9000230003500018000450005.60004.50005.800024000210003100074000210007.30003.3000110005300037000370002.000026000420007.80003.5000140005.10001.9000380003.00002.20003.50004.00002700013000210003.5000220007.20004.8000130007.50005.400030000196k160003.80001700026000280003800017000200006.90005900067000610001.60003.70004.90003000040000100001000004200095000280004.20003.6000170004.60007.90004.30003.60005.9000350001.80009.50001900025000370003.6000170002.90003.40002.90003.5000370001.500047000110009.100032000160002.50006.70001.90002700079k7.60008.2000330003.1000270001.50004.500043000210004.40001600013000340008.30002.50003.50006.40004.30006.30008.60001.9000118k490003.70003.5000530008.00002700024000110003.20003.70001.20008.4000310009.6000450005600026000170006.9000990004700036000210003.00008.50002.00001500050000170001.90007000041000990009600016000840005.00005.0000990009900026000370009.000088000870007800067000590001.400059000470007600062000800003.800056000720002500079000980006800091000440009600061000190008.8000960007.7000100002.2000160005.700029000240004.3000480002.3000130007.6000990002200012000880009100099000100005700018000330008.20004600063000239k4800048000180008.300054000300008.300023000110008.4000820005.0000170003700045000400007.00003.800022000150008.5000371k5.200013000920002.00005.00001.80001.700024000230005.10007.90007.80004.2000180003.30007.70003.100027000220k260003.000028000110006.1000260005.00002.10006.10002.2000140002.3000280003.00009.70003.40003.90009.400026000270004.4000260003.20007.00001.40001.70001.5000390006.10004.30008.3000730k2.100026000756k3.20004.1000200003.60004400026000170001.10002.900024000140006.90001200091k9.40001.7000293k350003.800017k74k200001.10004.10001300014k8.30009.00009.00003.30002800019000590001.9000120001.80003000018000290002600012000120001700018000260001700012000317k7.5000660003000035000260001500027000390009.100013000350005.10006.7000370003.90004.00007.60006.6000320002.900078k1.2000150001800015000190004.10005.60003.50006.40003.0000924k110004.90006.1000902k530001.30004.70003.1000110003.7000266k190002.00004.10003.400013000460003.9000110003.500011000170001700011000640001000035000120002.40003.70006.30002200010.00003.70004.600044000180009.70002.700051000470005600024000520004.90001.7000160006.500027000360005.2000430008.60002.0000140001500012000450003.2000620001500013000180001.7000150002.90005.90004.2000740001.7000470006.90008.7000818k280002900059000290002900059000270002900029000140001000081k280009.80002900028000318k5.4000280002900029000290003.0000270002900086000220002900029000290001600027000290005.70005.7000290002.000029000200002800027000290009.10002900029000350005.90001.200039000270003.20004.2000310003.900025000690004.30009.800013000160004.2000990002.30003.6000260003.20002.80008.30003.80003.40003.50001.90003.30007.10007.90001.40001.80007.50004.30002.80007.90003.700048000140005.7000939k5800042000350004300063000600002.80003.20004.200032000220007.30005.3000500007.500026000170004.2000310007.60001.7000169k630007.700045k6.600014000220003.60008.400025000475k4.3000350003.600024000350009100027000310002.70003200027000670008.2000300003.90001.10004800014000180002.90004.400089000130003.8000150001500025000340006.8000240003.4000240005.80004.900016000460003300043000560002.80001100026000110008.80006.20007.60004.90005.10004.20007.0000380008.70002.1000120005800018000200008.00003.0000120008.000028000290004.70004000014000160009.7000490003.7000270003.60005.30002.40008.50006.70009.0000266k27000670002200011000680004000015000230004.20002.9000850001.80001.1000220008.70002.60004.70002.900022000965k1.80002.00002.7000560002.70002.90004.900020000120003.80004.20001.4000210004100057000150005.100031000280003900034000310007600040000460008.70008000024000210009.70001200057000810006.70003.80005.9000460003.3000480003300042000990004.5000280008.1000210005600025000260009900025000990004300017000800004400012000100000690009000088000640003900054000560001400075000130008.100020000570003100039000290003.60003.80004.6000545k2.100061k130002.80009.90002.50005.4000140001.30002.5000283k2.30001.6000250005.40001.50005.30002.00005.4000410005.40002.40005.40009.0000250001.2000450005.30003.200073000110005.4000655k5.6000170009.50007.40005.40002.600096000730005.400035000480003700017000350002.300016000260002.9000714k12000520003.200049000280001200017000370002.000082000370004800054000460005100048000550002500033000100000700009.200043000910002900044000300009900090000720004700023000260005.0000200003000037000280008500022000820006500043000800002100093k9000067000320002.7000360003500041000710007000022000140006.000053000872k1.30001.6000560003.40003.1000200006.2000121k950009.40002.500093000700001.700037000700003.30001600019000220006800034000460009200045000500009.5000360007800022000480001.5000400006.5000510001200050000400005400045000630004.0000690003900055000460009900091000210004600045000240005200050000630005.80002300023000260008.8000270002.1000120005.0000230008.2000140008.50004700012000170003.900010.00004.1000150003.30006.7000110004.10003.70004.300031000250002.00005.80003.70001.70001200014000780006.400010000440001.5000850009.80005.70003.70004600030000470002300091000440004.80009.5000190007.30001.80005.70007.30004.00003.7000260005.60007.3000490008.700049000200007.3000322k6.30003.400022000250006.3000820008.70001000024000430003.7000130003.9000130003.00003.0000460003.00001700063000170002.90002.9000260002.7000240004.70002.70001.70003.00003.000031000360004.3000630008.600057000130005.60001.0000230001.6000240005.9000440003000076000976k1000044000190003.00001500015000430002.6000172k2400031000250002.3000330009.00007.4000190002.3000420003.30009.10002.800013000300003.00002.200014000238k37000520004.400028000320003.400023000610001500064000120004.40003.60002.9000100007.20003.8000549k3.90002.9000420001.10009.70001.700013000550006.50001.90005.1000360001100014000170004.30004.00004.00003.00001.2000290003.40005.10002.70009700019000100000730001000001800018000440005.20008.900034000300001.60004.40003.7000110004.70003300013000140003.70007.30003700013000130008.4000450007.90004.2000110006.20003.70006.00008.80002200030000110004.10003.400013000390005.10004.70005.50005.30002.80009.000051000206k954k260009.40002.40005.8000200001700035000170002.9000420005.1000160004.7000350001200026000170006000025000380002.00002.50009.2000444k15000120004.4000140006.70003.70004.00002.2000330005.70008700030000680005.90001900040000490005200063000240006.0000170004.200032000717k4.30007.20005.00001.2000310001700046000200008.30005.0000220003.10002.20002.0000120005.5000140001500081000620008.90003.000013000110003.400012000290002.60001.70005.4000300003100014000370004.00006.900014000210k130001800013000380003.80003.20003.2000609k1.70004.00006.5000140002.8000330003.30003.9000150009.50007.30002.40003.00003.50002.4000130005.30006.90002.3000150007.2000130007.30003.30007.30007.30002.6000308k4.5000210005.30002.300022000400002100021000530001.20005300011000270006.80007.40003.50001.70004.00008.8000380001.70006.100027000380002.80006.50003.6000530003.9000120008900046000540009.20002400041000705k250006200027000370003.70004.60009.90009.0000290002.50006.3000580004.10009.60003.800013000510002.7000820005.0000830002.50002.50006.60002.50003.70004.600080000110002.60002.5000240001200010000306k260002.80008.10003.20003.30001.60005.60004.5000110002.80002.80002.8000170002.9000190004.10008.40005.00003.90002.80002.60006.10002.60008.90009.10002.90002.00002.80002.80009.40007.80008.6000100005.10007.80005.90001.70002.30002.50002.80004.10006.2000200004.8000240002.4000130001.80009.800029000110006.4000870007.8000860002.10004.00001.90002.40002.9000210008.70001.70005.1000110001.30009.2000350009.20004.90001500015000190002.20002.10008.00003.50001600063000100007.5000210001.800011000180003.2000330008.400031000220003.60001500018000100002.500016000150004.30004.6000180008.7000220001.10004.80002.60005.20006.000043000904k200002800020000160008.70006.00001.80003.70003.2000160007.40003.30003.5000140008.900032000140003.80004.70009.9000170009.10001000027000250003300030000280007.300023000201k2.70007.7000210003.50003.300031000120008.10001400011000180002.9000473k760001800054000300002.1000480009200015000300002.600017000350001100065000250004.80004.80001700024000260004.8000175k150004.90008.8000360008.7000230005.10002300026000530003.2000230001.80005.40008.00005.400059000210005.1000550003.4000330004.70001.5000300002600054000100003.60005.200025000350k510005.10004.6000390001300031000190003.100017000383k2.40003.20002.10003000014000120003.00004.80002.80004.0000110001.80003.100020000330001700028000454k3.10003.20006.800099000170001200025000130002.10005.90003.70004.50004.1000110003.90003.00001.400011000120003000025000520006.30004.700025000620006.40001500070000120007.100047000230001000040000530008.20008.50003300034000230002.50001.80001.4000240002.0000421k1.2000110007.00002.10001600010000280005.5000410002.700093000400003.40009.00004.30002.90006.90001.00001.1000390003100014000290004.20005.30001.70006.3000270003.10002.50005.00006.30007.80004.80009.200034000470008.10002.500036000100006.2000100004.400018000160009.9000220001.9000260009.30002800010000110002.30004.40008.0000100003.7000210004.6000450004.20007.60008.60001500070k17000812k13000442k2.0000842k2.4000417k5.6000412k459k34000478k100006.2000335k3.8000782k721k430k429k192k200k130002.50008.9000417k460k5.9000728k496k816k6.3000414k334k506k2.00002.3000340007.500060000600007.9000240009500059000230001200011000140002.2000300008.5000190001700032000330004.200033000960004.70002600023000300006.4000280004900016000510002.6000250002700023000310002500016000704k7.40005.00002.600045000230008.40008.30007.9000260006.00004.80004.6000160005.70005.9000335k2.2000887k613k1.50001200026000130003.30002.9000100003.20002.50001.80002.70007.0000120009.6000180001.9000290009.40006.70001.60002.00002.50007.20006.30004.00001.20005.60003.50009.20001.200018000360001.20008.80004.10002.50002.700033000280006.7000210003.50006.000022000270003.40001600012000100009.000036000120003.60003000019000130003300014000243k240006.600044000890003.50003200050000150003.10003000034000130002.60002.50007.5000569k260005.20005.40009.100026000280003.800092000206k1900093000130001.8000100003.50007.500068000778k2.10009.20003.00007200035000310004.10001300060000850001.30006.90005.50005800094000683k592k6.40003.3000319k2400036000186k3.5000840k410001.500036000647k5.3000190002.5000930002.80005.60007200021000191k2200067000842k1.10006.9000150001.10006.20001.4000520008.7000373k3.80003.80003.80008.2000437k5.70001.6000160002.60006.6000598k2.7000716k140001200046000130001.90001.20004.3000130008.40007.3000585k3.20006.50002.70001.4000170003.10002.40001.6000982k26000190002.90004.5000222k420007.9000219k55k948k7.3000323k890005.200015000190007.60005.600018000240001000011000290004.300010000880005.30005.40003.10006.0000260006.50001900033000400009.20002800021000200004.0000260002600033000250004400019000270001.5000130004.80001.60004.90003.50008.70007.7000190003.40002.90002.40002.70005.00005.90002.80003.400012000200004.200010000260004.00003.6000980005.70003.6000960002.2000100004.80002.0000620009.800034000530007.30002.8000240007300053000270004.4000170001.3000190008.80001.30001.80004.80001.8000691k12000200005.10007.2000240007.4000530001.3000140008.80001.7000140001300017000240001700026000511k1.70003200012000230005.50009.4000180001.8000951k1.90002.60001300016000963k8.5000230008.20003.80002.40001.7000100002.000025k4.6000310008.2000554k210006.90002100011000340004.10001.90003.50003.00004.00002.2000180003.70002.40009.3000440003100048000380004900027000351k8.800023000480009.700027k120007400044000360001.400025000440003600082k570009.80002400047000370006.900010000150009.400049000240007.30002400014000210008.80004.0000180001800013000130002400025000430004.600047000200002600069000130009.200019000340001300034000240003.80008.30002500053000240005.40007.7000110005.60005.4000170004.40008.200098000290002.40008.50003.20003.800033000530006.2000220001700025000390004.60001300036000130008.500025000310003.70008.500024000340006.90004500012000330003500025000480001.00007.2000140008.9000208k9.40008.30006.9000120005.500026000980005.20007.70005.3000375k3.70002.9000180006.0000340007.0000380005.40001.20001.20007.90002.30002.3000323k1.40002.500012000150003.1000120001.50009.00003.70008.7000190007.40003.3000913k4.90007.80005.00001200011000250002.60007.20001.80002.6000140002.20001.90005.6000514k2000011000120005.60003.80006.40003.70003.40008.40002.90002.90001.80003.1000800003.300015000360008.30003.10009.50004.80006.70001.50009.20003.4000130005.800020000200006300013000270003.8000160002.3000120005.600037000220001.70002.60007.9000200005.7000160001.70004.20004.60002.70007.20004600014000120003.70001.60001200029000360004.000015000420007.80004.60004.30002.8000270007.300023000200005.400024000200001.40003.60006.30002100014000250001600055000220003.00002.600024000551k4.00005.100022000210001800062000300002300010000290001.900029k430008.300019000103k6.20008.30005.500011000170002.40004.60003.0000210002.7000150004.70003.00001.400013000160002.0000110002.900040000140009.70007.80003.20008.0000290003.10003.5000140008.70008.80009.70003.10003.20008.50003.50007.8000110006.3000190005.0000160003.50002.30003.30004.10004.10004.10007.00002.30004.10001.1000180004.10001.50002.30003.50003.70004.00001.80004.00003.8000100001.30006.60004.10002.50004.10002.50002.50004.30007.70006.80003.40001.40003.60002.300014000990004.10008.70002.50002.90004.30006.100014000150006.3000660003.4000760007.70006.500014000250001.4000500002.50002.20004.600015000130008.1000160009.6000100003.10007800027000100005.80003.2000898k3.70005700042000360001.200046000300003.2000100001.8000310006.60005.10003.20006.00002.50009.900010000172k5.20005.10006.400014000150003.30006.90005.10002000010.00006.60006.10003.20001000001.40002.10007.2000910002.8000743k1400014000430006.900096000140002600018000340002.80009200011000150004.00003.80001.3000130006700099000116k2.200024000430005.90002.40006700067000130008.20005.20009.2000630004.20001600010000180001200078000380009.500045000280007.70009.50005.2000380003.70002500020000153k5.30006.300038000130006.40001.90006.40002.00008.6000240001000039000600004000015000209k1.50002.3000130009.40003.70004.200022000130008.100012000190004.90009.70003.60002.50002.500018000350006.30001.3000240003.40003.40003.40003.4000170002.20005.00004.60009.00001.3000240003.200097000353k3.20002.40008.4000120003.40003.40007.30004.2000580006.8000499k3.20003.30003.700025000120001.3000400005.200017000173k5.10005.80001.20005.80001.80001.90003.60002.50002.10001.1000170002.10002.80005.50004100014000597k2.70003.2000809k2.20008.20002.80004.00001.800070k2.30001.3000122k4.1000110004.20008.2000411k2.30005.70002.70001.9000190002.300049000280004100025000250003800028000550005800027000250003100094000250002100045000450002400083000880005100063000210007.20005100098000270009.30008700055000260006200022000400005.500018000740001600064000160002500012000180004.5000770008500085000260002.00003.900011000180008.70003.90004.0000530004.900052000230001300030000180004.100013000200002100012000180004100012000140006.000017000400k42000540001100037000130006.500016000190003.3000350006.3000120005.0000430002.90004.10008.50007.0000190001600014000801k5.6000360003.20007.00002.900035000210005.80004.7000200003.9000120007.800024000120004.90003.30005.40002.400027000220001.20002.5000140006.80001.2000787k4.90005.30006.500069000140003.4000500008.10001.70001800018000360008.10005.10001.6000610004.100010000237k3.4000180003.50003.40009.100015000140003.800010000260003.300015000190003.70002.100011000430006.900040000360001.500037000220003.30002.6000340002.9000120003.70001700023000170005.0000180003300026000320003800021000410001.70004.30009.90005.90002.70001200016000190002.80003.6000370004.30001.70003.1000300005.20002.8000180003.400041000130006.70009.80001400026000220008.70008800018000180001.30004.50007.80002300050k7000047000420003.700039000220002200012000140004.30008.50006.70007.5000880001.10008.30005.7000643k440001100028000400009.10002.1000320007.80002.4000530006.0000250003.3000340003.60006.200068000190003.60001700031000180006.000029000770004200014000290005.7000360008.90005.1000190004.300062000290007.800020000986k18000140009.3000170001.4000160004.00008.30009.60001100020000140001700015000130001100046000110001.600028000110004.90008.600023000240001.60008.400097k2000044000880004.80003200025000270007.300014000150001600025000140002.9000180003500025000140006.50005.500028000170007.3000240002.1000240002.80007.20008.60002.30001200070000340002.60003500024000880007.00003.90003.8000260005.00004.70005.60005.30002.2000260005.6000260008.6000210004.80001500020000200005.50005.7000110004.70004.10006.9000120005.90002.20007.10007.900025000260004.9000250008.30006.00003.0000190004.10006.00004.6000260004.4000100005.000016000220008.100021000130004.80006.10004.30006.20003.50002.4000380009.5000516k837k2.0000170006.80001.1000780k3.7000961k2.60004.80007.8000220001.50009.60002.400055000269k3.700021000960001.100020k12000498k130004.200011000600k180003.000014000720005400012000370006.50006.6000500004.30007.60005.5000940003500055000110001200021000270005.30003000010000110009.8000680003.8000330002400035000200003.40002.5000320005.80001800020000320006.90001200017000592k5000019000550001.00002.00007.500016000100001.1000810001500013000200007.10004.50001.30001000029000749k200008.10003.80001.60006.10004.900020000150009.3000120003.9000610005.80005.00004.4000642k9.40003.30006.40004.00004600012000300001.4000150001.8000190001200021000250003.40008.7000190009.800018000180001500017000430009.3000110002.00002.20004.40005.4000120001.400017000881k3.300017000130007.3000230001200072k2.20003.5000180008.60004.80007.3000130005.30006.2000656k5.0000601k8.3000730007.40002200026000140007.30003.10001100012000130003300049000310003400044000110005500059000600002500056000200002.3000221k1.500048000310007.5000228k2.60002.30002.8000160002.20004.500033000720003.30006.60005.20008.0000108k3.500011000100009.400010000940k8.70006.30004.1000370003.2000110002.60003.000014000470002.50002.300049000440003.00002.300015000100007.60005900036000320009.5000950005.800035000630009.700091000630003.10008800012000920004.1000230007.20006.40003.4000910006.1000220006.90009100072000100002.20008.50001200021000110002.30001600032000170001.300018000100001.20008.700042000380005.30005.8000960002100010000800006.10001.50003100026000200002.9000380003600014000176k44000820003.3000180008.300033k140008.30009.90008.00001.50003.6000150006.7000663k1.4000200006.0000140001.50002.3000270007600015000320006.3000240001.400036000150003.00001400010000150005.20002.90003.30005.7000360005.70002.90003.30004.50001.60002.80001.10004.10001.10006.8000280003200012000220001.40001.2000200007.60002.90002.50001100034000210004.50002200034k14000120001.30003.1000110005.90003.0000120001.1000260004.50005.90005.7000370007700043000200008.000017000100003400081000350001000004800059000140001000016000240002.30004.10002800027000270006.7000230004.90008500041000140009400045000942k26000990005.900036000520006300026000350005.4000530004.00005100020000390005.00005.700029000870007.1000610003000052000440004.6000900009100030000200002800016000160007.9000570001600025000340002900087000380001800019000700003.300014000660004.300048000310003.3000250001100017000160003.5000330006.40003.90001.50006.6000230002.4000350001.10001100013000340007.10004.2000170003.00005.3000300003.30002.50002.00006.30002.30007.20002.20006.20001.7000100003.200030000110002.40001.70003.3000130003.20007.90005.2000220001.60009.40005.40001.10003.30005.50004.2000100005.00002.9000300005.40003.70002.90003.00004.2000110003.00006.30002000097000259k4.90001100036000164k3.000016000458k245k5.20005.70003.300027000100007.80008.50005.9000629k1.700092000210001.200033000350002.400011000230004.0000170004.100028k1.80003.400032000288k4.2000170001800037000150003.8000210004.2000280002.3000775k100005.9000330007.5000190003.80004.0000620007.6000270002.1000150003.400030000180005.30003100050000220002.60008600071000320009.2000160004.20003.5000600005.900078000120004.90003.00006.30008.40004.00001.9000201k5.00004.0000300009.00008.5000130003.4000200004.40008.3000190006.40002.40005.30005.200070000410001.9000390005.8000110006.9000150004.70008.2000180004.4000150009.40003.0000120003.40005.10002.00003.20005.200011000140007.7000120003.80008.10005.40005.7000100008.70003.80005.60003.80004.1000110005.70003.80009.10002.50006.7000110007.70007.0000430001300020000590003.90002700016000630001200030000440004300017000420006.30002.60002000016000130006.30002.7000560001.10002600056000785k250004.70006.00004.900018000660004100096000760004.2000360003.900024000350002300013000190007.60009600035000810009500010000210004.700027000250003.4000330005.70003.10004.30004.50004600033000580009.30006.20006.7000410006.00009.50006000031000360004.30001.5000680001.20001200022000450006200080000340003.500067000580005.10005.50001200019000190008.80001100012000320002200038000600003500015000430004.40002.1000230002.3000636k4.20004.60007.90001.400021000916k280001.6000840006.60001200014000380005.00002.90001.40008.40003.00002.60003.70001000005.8000280004.50008.90003.1000220009.9000490001.60008.90002200023000940006.30001.100095000210009.5000240001.5000240007.90005.30001.2000120005.70001300024000540003.900051000490001.3000994k3.60004.100066000320002.90008.9000470002.1000120007.8000150002500014000100008.90004.80003.20004.70002.0000140004.700022000140004.30005.2000309k8.7000485k2.50009.200039000960006.90009.8000160006.2000914k8.200013000480002.8000310001.30002900017000190002.6000710002.600014000440005.100011000130002.6000650008.7000903k250004.00001.70002.40008.5000260002.00002.5000180002.90001.30001.70002.7000608k220004.2000410006500016000150009.30004.20005.10008.9000240004.0000310009.50008.2000170001.800030000310008.50006.3000100004.80006.00005.80006.3000790006.30002.40005.00002.00004.40008.0000390003.50003.90009.80005.9000150006.30003.00002.20004.2000160001.5000500k3.40003.3000130002.5000100003.000054k4.80002.90003.000040000110006.6000160001.6000480004.60002.800023000260002.500024000560004.00002.5000400007300031000450004200039000480007.20001000014000110002100032000470002.100016000330003.4000170001.60004200011000480004000020000600002.6000220003.20005.30002200010000430004.8000200001.90003.30001.20005400026000670006.4000500002.50001.3000210007.4000300003.90009.80004.50009.80003.2000470006.00002.50002.3000140007.0000562k140004.70004.5000100003.00001800032000847k2.80009.9000170002.50003.40006.8000957k290003.6000280003.10002.50001.600023000150006.1000100002.40004.60003.3000100003.80003.1000330006.20002.100018000370005800022000710006300049000150006700045000290005700043000120009600026000790009800019000280005000039000300009900058000920004300010000072000370003.1000330003.30005.6000160005.40001100030000310003.3000300006.10005.20001800016000250002.3000130008.8000150002.4000700003.90003.90009.0000210007.30001.80001.4000948k5.00002.2000688k2.20002.50007.3000900003.1000811k4.00004.10002.8000160004.30003.8000270k3.70003.00002.30003.80002.70002.2000280001.50005.50004.000048k7.7000260008.60004.3000170003.8000150003.40005.00002.30001.60002.70008.10003.80006.30007.4000120007.90005.40004.00003.10005.50008.40003.9000329k3.00004.20003.80005.30006.00004.3000300004.600017000140003900013000350006.900014000230001500011000140001400017000170009.8000150004.40004300019000360009.400029000100002300010000330002000033000140002.3000540002.50004.300011000150001.60002.0000310003.60003.7000140006.60005.10004.60005.30006.8000530007.5000230001.50002.40002500030000130006.50004.4000230005.0000160007.40001.10001.100038000523k118k180001600061000560003.1000120006.00004.00003000017000510003100042000990006.4000220006300011000990002100014000460001800061000990003100026000260007.00003.30003.90008.7000620003.50006.4000921k7.1000874k2.30005500041000520003200013000440003.40009400023000480002600019000440003300035000650009.2000230004800037000200004.80005.70002900042000480003.70006100066000820001.60001.800085000981k120004.500014000150003.8000784k7.7000130001.5000130007.10003.200021000230002700019000450001500013000330002.1000160003.3000200005.20008.700099000770003.80006.00005.0000550005.3000370002.30001200050000280k310001900069000250003800026000130006.80004.10004800068000450009.30004400041000650004.40004600023000300002800047000590001.600014000430004.10003.80004700024000310002.70007.300059000480001.3000120005.00006.6000550002.800010000300003.7000180007.4000170001.30003.7000260005.00003.30003.9000290005.60003.00001.80005.80005.70002.6000380006.2000110001.5000620004.00003.100056000300001.90002.00004.50004.30009.90002.00002.300029000990005.10002.700064000490007900047000250004900036000220007.000060000130003.200031000500006000016000250003.60003.50001200033000240004.1000490003.200073000220004.4000320006300015000160006.40002500052000210002.3000150006.800040000880005900024k46000390006.90001400052000140003700024000280004.40005200051000760004.4000260006700046000780003300026000600003.00009000028000600004000025000410001300042000350001800053000310004.3000110008.50002100043000100003.4000150007.1000110001.50002100034000110008.9000190004.20001.100011000140003000024000518k6.300041000180005.20006.70003.10002.0000360006000074000110009200083000540004200023000310009800069000754k37000200007700073000330002.7000570008300073000440002.8000130002.1000280002.3000892k700002.3000200008.8000240002100015000154k210004.800020000310001600012000270006.3000110002.00001.500020000390003.2000270007.200023000470003100021000340005.3000370005.7000100005200069000560005300062000760007100072000280002.800051000280004000027000370004600050000490005800041000730003300046000190003200030000920005500045000470008500048000770003000028000350005.80003900025000320008.900084000600005.30003900020000200001700018k42000290006.60003.200033k4.40003.4000860k1.20001.70002.8000364k2.20001.90001.70003.0000387k2.70002.00005.0000375k11000626k110002.90006.50002.00004.20006.70009.20004.700015000161k240001.70002.70005.90009.2000130009.20006.30003.80004.20005.7000180009.9000120005.30009.9000630001300031000350006100062000930009300079k5.0000330008.7000610006.10002600037000190002.30004.400032000879k5.3000170004.20005.30003.4000120001.8000120002.50003.90002.00007.2000130002000016000220003.50002.300065000210002100020000120002.00002.7000130007.00009.4000110001.2000180007.50002.6000120003.100039k4.300045000150002.50005.40005.20001.5000970k1.40002.0000340002.10009.70002.20001.60001000042000130002.70008.700032000100006700017000400002100015000120009.7000310004.00008.600025000260006.4000880005.8000410002.50006.6000240009.80003200072000190008.10006.20008.4000190008.5000430002300048000200005.20003.9000520001.30002500011000160001.000061000150005400019000190003300064000720004500027000690004.90004.40008.6000130008.700012000290005.80003600017000170001.9000120007.100096000210006.5000620008200069000220003400063000140001.80009500030000160004.300067000120008.50004800062000950005.2000230002.8000130001.10004.6000170k1.10001.000011000141k160k86000144k141k143k6.3000170007.9000190k6.00003.50002600093000210002.3000640008.1000376k130008.000029000420001900017k2.50003.60003.9000540006.80003800044000300008.4000270001000015000270005.10003.40001000019000540003.10002.70001.70008.40001.2000190001200044000260009900019000180003.300030000240004.5000150001000018000130004.50003.3000210001000036000330004.30009.6000656k61000140005.40001.30002.8000470007.60004.600010.00004.6000130001.20002.9000100004.8000360003.50005000051k110002900046000460001.40001.2000228k9.800039000193k12000192k570002.10003.9000110002.30008.50001.4000473k2.40003.9000680001.60001.80004.00001.5000260003.000071000246k3.60001.300024000680001.3000440007.00005.40007.600014000710002.60001000048000250006.30009600010000470003.20005.8000230001.10004.60006.30004.30004.80004.20003.60004.00003.90008.70009.90005.10001.80001.50007.0000180003.40003.30001.50005.1000100001.60003.20003.0000100002.80002.50001.300011000150003.10002.80003.70006.10004.5000300006.20002000038000200003.10006.40007100011000110002.600033000210004100012000210002.80003.60001400029000760005.700017000280004.6000190003.10007900036000170007.80008900049000120003.00005700016000130005.9000400001400073k3.5000170007.9000190003000012000420007.40009.10007.50004.80007.40004.90009.10004.20002.7000750003.90002.3000350003.50009.10001.5000260007.40002.10004.70003.40004.20004.60007.1000210002.70004.400014000170009.00004.50002100015000210008.00001100053000210002.9000200002.9000200004700012000120001700010000260009.20009.800017000658k300001.8000970002.9000150005.30008.3000200003.00005.60007.9000220005.000011000150007.50004.1000130001200060000250004.60001.70006.500036000170006.50003900046000520007.7000300001700039000300001200045000290004800050000660002700051000690002100014000190006100028000490009400023000940002.500013000340003300016000992k180008.1000150008.20008.1000270004.1000200006.0000170004.7000140005.80009.90009.900049000140007.60007.800046000290003.600049000290008.2000540001800056000260003.30004000014000240003700022000160002.00003.50008.5000253k957k420k130002.40003.40005.10009.700072k2.4000350003.900026k29k5.80002.80005.80002.10001.40003.90007.5000580004.0000404k4.1000140002.60007.6000110008.00002.300049000430001000016000240006.200027000460002.400036000170008.00003700033000150001500013000110002500016000150001500043000430003600023000990004600020000450002.90002100035000410002.20002.5000350001.50005400044000270008.500022000260006300020000260003.500026000100005.5000260002600026000260002600026000260002.400026000260002600026000260002600013000260002500026000260001500018000210002600017000260001200026000260002600026000260002.6000270001500026000170004500024000110003.1000300003.20005.60007.10007.100031000340009.200036000150002.00006.1000220004.4000860001300012000410008.20008.000015000470k5.30009.40002.200043000226k200002100037000260003.4000220007.80003.4000280007.80002.60003.90007.60008.10009.50001.8000160003.200045000140001200049000130009.90003.90007.6000280001200069000340009.20005.40008.7000180005.700012000150001.600038000110001400030000130001900010000450005.3000190007.00005.600011000120009.200011000130003.3000140008.80007.3000140008.50003.30009.0000240k8.20007.90001.6000340002.1000590005.200089k3.6000234k257k9.80003600020000280003.20005000041000364k3.1000861k1.60001500017000160001500011000160001500019000160002.80001.300021000180005.20002.00009.7000170004.3000540007.20001.4000110004900015000250004.00001600011000400003.700024000160008.10006.30006000046000467k1.400022000157k2.60004400011000390004.40001.70007.90001.20002.00005.8000610003.80003.3000290003.300044k26000120003.3000676k2.5000720004.300067k2.40008.000011000552k885k7.0000160002400012000410002.40003.90008.9000360009.0000240002.2000380007500050000440001100072000840009.50002.8000480002000048000200003800016000780005.70004.00007.8000460001020k6.8000120001900028000810001700015000420004.20001.000024000210003.9000130002.7000310004.90006.80008.00001.50003.60008.60002.50003.10002.9000820007.700013000130007.40002.30009.8000582k619k2.60009.6000530003.60009.500019000 to numeric

In [49]:
df['Size'] = df['Size'].astype('float')

ValueError: could not convert string to float: ''

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          10840 non-null  float64
 3   Reviews         10840 non-null  int32  
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), int32(1), object(11)
memory usage: 1.1+ MB


In [61]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews   Size  Installs  Type Price Content Rating  \
0      159  19000   10,000+  Free     0       Everyone   
1      967  14000  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [62]:
df['Installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

In [63]:
df['Installs'].isnull().sum()

0

#### None of the data are found missing in Installs column

In [64]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews   Size  Installs  Type Price Content Rating  \
0      159  19000   10,000+  Free     0       Everyone   
1      967  14000  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [65]:
df['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

In [66]:
df['Price'].isnull().sum()

0

#### None of the data in the price column is found to be missing

#### Cleaning the Price column

In [68]:
chars_to_remove = ['+', ',', '$']
cols_to_clean = ['Installs', 'Price']
for item in chars_to_remove:
    for col in cols_to_clean:
        df[col] = df[col].str.replace(item,'')

C:\Users\Dell\AppData\Local\Temp\ipykernel_13572\4019055661.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df[col] = df[col].str.replace(item,'')


In [72]:
df[['Installs', 'Price']]

Installs Price
0         10000     0
1        500000     0
2       5000000     0
3      50000000     0
4        100000     0
...         ...   ...
10836      5000     0
10837       100     0
10838      1000     0
10839      1000     0
10840  10000000     0

[10840 rows x 2 columns]

In [73]:
type(df[['Installs', 'Price']])

pandas.core.frame.DataFrame

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          10840 non-null  float64
 3   Reviews         10840 non-null  int32  
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), int32(1), object(11)
memory usage: 1.1+ MB


In [75]:
# Convert Installs and Price columns to float type
col_list = ['Installs', 'Price']
for i in col_list:
    df[i] = df[i].astype(float)

In [76]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          10840 non-null  float64
 3   Reviews         10840 non-null  int32  
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  float64
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(3), int32(1), object(9)
memory usage: 1.1+ MB


In [77]:
df['Type'].unique()

array(['Free', 'Paid', nan], dtype=object)

In [79]:
df.columns[0]

'App'

In [85]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [80]:
df.columns[10]

'Last Updated'

In [81]:
type(df.columns[0])

str

In [87]:
for col in df.columns:
    df.columns[col] =  df.columns[col].str.replace(" ", "_")

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [92]:
df.columns[10] = df.columns[10].replace(" ", "_", )

TypeError: Index does not support mutable operations

In [91]:
df.columns[10]

'Last Updated'